Project

Nessing opened the pull request:
Создан пользовательский интерфейс, отправка запросов на сервер через кнопки и возвращение ответа с сервера клиенту о результате.

=== Дополнительная информация по классам ===

Server.java - запуск сервера;
ServerHandler.java - логина по обработке запросов от клиента;
Client.java - запуск клиента (бэк часть);
StartGuiClient.java - класс для запуска пользовательского интерфейса (на JDK 11 необходимо запускать класс,
который наследуется от класса Application в сторонем классе);
ClientAuthorization.java - логика по работе клиента и взаимодействию с сервером через GUI;
ClientControllerAuthorization.java - класс для взаимодействия с окном "Авторизация";
ClientControllerSynchronization.java - класс для взаимодействия с окном "Синхронизация";

Nessing committed:
result_1

Создан пользовательский интерфейс, отправка запросов на сервер через кнопки и возвращение ответа с сервера клиенту о результате.
=== не реализовано ===
передача данных через json и подключение БД для авторизации и регистрации клиентов

=== Дополнительная информация по классам ===

Server.java - запуск сервера;
ServerHandler.java - логина по обработке запросов от клиента;
Client.java - запуск клиента (бэк часть);
StartGuiClient.java - класс для запуска пользовательского интерфейса (на JDK 11 необходимо запускать класс,
который наследуется от класса Application в сторонем классе);
ClientAuthorization.java - логика по работе клиента и взаимодействию с сервером через GUI;
ClientControllerAuthorization.java - класс для взаимодействия с окном "Авторизация";
ClientControllerSynchronization.java - класс для взаимодействия с окном "Синхронизация";

Nessing committed:
edit ServerHandler.java

Nessing committed:
delete useless class

Nessing committed:
add test code

добавлен код для отправки файлов

Lezenford commented on src/main/java/client/Client.java:
```diff
@@ -0,0 +1,79 @@
+                    nioSocketChannel.pipeline().addLast(
+                            new StringEncoder(),
+                            new StringDecoder(),
+                            new ByteArrayEncoder(),
```
странная очередность в пайплайне. стринг декодер принимает массив байт и преобразует в строку. дальше у вас попытка работать с байтами. убедитесь, что эти элементы там точно нужны

Lezenford commented on src/main/java/server/ServerHandler.java:
```diff
@@ -0,0 +1,182 @@
+                while ((s = fileInputStream.read(bytes)) > 0) {
+                    count++;
+                    System.out.println("== FRAME == " + count + " " + bytes.length);
+                    stringOut = new String(bytes, "ISO-8859-1");
```
обсуждали, что при таком подходе часть байтов может быть от предыдущего сообщения, если на чтение в конце файла осталось менее 8192 байт

Nessing committed:
Step_2

добавлено скачивание файла с сервера на клиент.

Nessing committed:
edit class ClientHandler and ServerHandler

Перенесена логика из ClientGui в класс ClientHandler.
Начата реализация отправления файла с клиента на сервер.

Nessing committed:
edit classes Client and Server (edit pipeline)

Lezenford commented on src/main/java/client/Client.java:
```diff
@@ -0,0 +1,80 @@
+                @Override
+                protected void initChannel(NioSocketChannel nioSocketChannel) {
+                    nioSocketChannel.pipeline().addLast(
+                            new LengthFieldBasedFrameDecoder(Integer.MAX_VALUE, 0, 4, 0, 4),
```
Вы уверены, что использовать MAX_VALUE хорошая идея? Это фреймы объемом 2 ГБ который хранятся в оперативной памяти. если для клиента это с натяжкой позволительная роскошь (хотя в целом нет), то для сервера это однозначно неподъемные значения при наличии хотя бы пары клиентов с большими файлами.
Так же не хватает класса Prepender который бы дописывал длину сообщения в отправляемом потоке.

Lezenford commented on src/main/java/client/gui/ClientGui.java:
```diff
@@ -0,0 +1,226 @@
+import java.util.logging.Logger;
+
+public class ClientGui extends Application {
+    Logger LOGGER;
```
не стоит использовать неприватные поля, тем более они хранят состояние

Lezenford commented on src/main/java/server/Server.java:
```diff
@@ -0,0 +1,71 @@
+                @Override
+                protected void initChannel(NioSocketChannel nioSocketChannel) {
+                    nioSocketChannel.pipeline().addLast(
+                            new LineBasedFrameDecoder(Integer.MAX_VALUE),
```
у вас на выходе стоит энкодер который пишет количество байт в пакет, а не входе стоит вариант который ждет окончания строки. Приведите в единый вид, сейчас скорее всего проект будет с ошибками

Nessing committed:
edit Server

Добавлена возможность принимать файлы сервером от клиента.

Lezenford commented on src/main/java/client/Client.java:
```diff
@@ -0,0 +1,82 @@
+                protected void initChannel(NioSocketChannel nioSocketChannel) {
+                    nioSocketChannel.pipeline().addLast(
+                            new LengthFieldBasedFrameDecoder(16384, 0, 4, 0, 4),
+                            new StringDecoder(),
```
я бы опустил StringEncoder на одну позицию вниз, сразу за препендером. Но 
это не критично в данном ситуации.

Lezenford commented on src/main/java/client/gui/ClientGui.java:
```diff
@@ -0,0 +1,240 @@
+
+    @Override
+    public void start(Stage stage) throws Exception {
+        LOGGER = LOGGER.getLogger(ClientGui.class.getName());
```
это можно вынести сразу в поле и сделать его статическим - обычная практика для логгера

Lezenford commented on src/main/java/server/Server.java:
```diff
@@ -0,0 +1,72 @@
+                protected void initChannel(NioSocketChannel nioSocketChannel) {
+                    nioSocketChannel.pipeline().addLast(
+                            new LengthFieldBasedFrameDecoder(16384, 0, 4, 0, 4),
+                            new StringDecoder(),
```
аналогично клиенту, я бы опустил декодер на одну позицию вниз. И так же не криитично

Lezenford commented on src/main/java/server/Server.java:
```diff
@@ -0,0 +1,72 @@
+                @Override
+                protected void initChannel(NioSocketChannel nioSocketChannel) {
+                    nioSocketChannel.pipeline().addLast(
+                            new LengthFieldBasedFrameDecoder(16384, 0, 4, 0, 4),
```
вы ограничили максимальный размер фрейма 16 килобайтами, но заложили под эту длину 4 байта длины. в 4 байта можно записать число обозначающее 2 ГБ. В данном случае это слегка оверпрайс.
Да и ограничение длины строки в 16 кб даст сложности при отправке файлов - придется дробить на очень большое количество кусочков. Возможно стоит увеличить максимальную длину?

Nessing committed:
edited options send and get package files

Добавил возможность получения комплекта файлов с сервера.
Проведен рефакторинг кода (добавлены методы на стороне сервера)

Nessing committed:
Add send and delete files on the server

Добавлена реализация отправки комплекта файлов на сервер и удаление с сервера тех файлов, которых нет на клиенте.

Nessing committed:
add class ServerGui

Добавлен класс ServerGui для отображения визуального окна при запуске сервера, а так же для указания пути к папке, где хранятся все файлы клиентов

Nessing committed:
add client and server treatment exeption

Добавлена обработка исключений.
на стороне сервера:
- проверка правильности пути (директории)
на стороне клиента:
- если соединение не установлено (автоматически при запуске приложения);
- если соединение было разорванно

Nessing committed:
edit pof file
